src/hardware/imf.cpp

```diff
@@ -0,0 +1,13422 @@
+	PACKET_STATE_36 = 0x36,
+	PACKET_STATE_37 = 0x37,
+	PACKET_STATE_38 = 0x38,
+	PACKET_STATE_39 = 0x39,
+	PACKET_STATE_3A = 0x3A,
+	PACKET_STATE_3B = 0x3B,
+	PACKET_STATE_3C = 0x3C,
+	PACKET_STATE_3D = 0x3D,
+	PACKET_STATE_3E = 0x3E,
+	PACKET_STATE_3F = 0x3F,
+	PACKET_STATE_40 = 0x40,
+};
+
+struct BufferFlags {
+	// bit-7:true=no data / false=has data / bit6:read errors encounted /
+	// bit5: overflow error / bit4: offline error
+	volatile uint8_t _unused1 : 4;
+	volatile uint8_t _hasOfflineError : 1;
+	volatile uint8_t _hasOverflowError : 1;
+	volatile uint8_t _hasReadError : 1;
+	volatile uint8_t _isBufferEmpty : 1;
+
+	inline void setDataAdded()
+	{
+		_isBufferEmpty = 0;
+	}
+	inline void setDataCleared()
+	{
+		_isBufferEmpty = 1;
+	}
+	inline void setReadErrorFlag()
+	{
+		_hasReadError = 1;
+	}
+	inline void setOverflowErrorFlag()
+	{
+		_hasOverflowError = 1;
+	}
+	inline void setOfflineErrorFlag()
+	{
+		_hasOfflineError = 1;
+	}
+	inline bool isEmpty() const
+	{
+		return _isBufferEmpty;
+	}
+	inline bool hasData()
+	{
+		return !isEmpty();
+	}
+	inline bool hasError() const
+	{
+		return _hasOfflineError || _hasOverflowError || _hasReadError;
+	}
+	inline bool hasReadError() const
+	{
+		return _hasReadError;
+	}
+	inline bool hasOverflowError() const
+	{
+		return _hasOverflowError;
+	}
+	inline bool hasOfflineError() const
+	{
+		return _hasOfflineError;
+	}
+	inline uint8_t getByteValue() const
+	{
+		return (_unused1 << 0) | (_hasOfflineError << 4) |
+		       (_hasOverflowError << 5) | (_hasReadError << 6) |
+		       (_isBufferEmpty << 7);
+	}
+};
+
+template <typename BufferDataType>
+struct CyclicBufferState {
+public:
+	CyclicBufferState<BufferDataType>(const CyclicBufferState<BufferDataType>&) = delete;
+	CyclicBufferState<BufferDataType>& operator=(
+	        const CyclicBufferState<BufferDataType>&) = delete;
+
+private:
+	const std::string m_name                    = {};
+	SDL_mutex* m_mutex                          = nullptr;
+	volatile bool m_locked                      = false;
+	volatile unsigned int lastReadByteIndex     = 0;
+	volatile unsigned int indexForNextWriteByte = 0;
+	BufferFlags flags                           = {};
+	volatile unsigned int m_bufferSize          = 0;
+	volatile bool m_debug                       = false;
+	// volatile BufferDataType m_buffer[2048]; // maximum that is used in
+	// the IMF code
+	volatile BufferDataType m_buffer[0x2000] = {}; // FIXME
+
+	void increaseLastReadByteIndex()
+	{
+		if (m_debug) {
+			IMF_LOG("%s - increaseLastReadByteIndex()", m_name.c_str());
+		}
+		lastReadByteIndex = (lastReadByteIndex + 1) % m_bufferSize;
+	}
+
+	void increaseIndexForNextWriteByte()
+	{
+		if (m_debug) {
+			IMF_LOG("%s - increaseIndexForNextWriteByte()",
+			        m_name.c_str());
+		}
+		indexForNextWriteByte = (indexForNextWriteByte + 1) % m_bufferSize;
+	}
+
+	inline void setDataAdded()
+	{
+		flags.setDataAdded();
+	}
+	inline void setDataCleared()
+	{
+		flags.setDataCleared();
+	}
```

Comment:
How about using UpperCamelCase for method names for consistency?

Reply:
(marking resolved because not renaming variables)